File: Solutions/Python/best_time_to_buy_and_sell_stock_iv.py
"""
    [HARD]
    188. Best Time to Buy and Sell Stock IV

    Concepts:
    - array
    - dynamic programming

    Stats:
        Runtime | 315 ms    [Beats 9.79%]
        Memory  | 84.21 MB  [Beats 7.91%]
"""

class Solution:
    def maxProfit(self, k: int, prices: List[int]) -> int:
        cache = defaultdict(int)
        n = len(prices)

        def transact(date: int, limit: int, is_buying: bool) -> int:
            nonlocal cache

            if not limit or date == n:
                return 0
            
            if (date, limit, is_buying) in cache:
                return cache[(date, limit, is_buying)]

            if is_buying:
                # either buy today or wait to buy another day
                cache[(date, limit, is_buying)] = max(
                    transact(date + 1, limit, False) - prices[date],
                    transact(date + 1, limit, True)
                )

            else:
                # either sell today or wait to sell another day
                cache[(date, limit, is_buying)] = max(
                    transact(date + 1, limit - 1, True) + prices[date],
                    transact(date + 1, limit, False)
                )

            return cache[(date, limit, is_buying)]
            
        return transact(0, k, True)

File: Solutions/Cpp/best_time_to_buy_and_sell_stock_iv.cpp
/*
    [HARD]
    188. Best Time to Buy and Sell Stock IV
    
    Concepts:
    - array
    - dynamic programming

    Stats:
        Runtime | 482 ms    [Beats 5.01%]
        Memory  | 80.83 MB  [Beats 6.21%1]
*/

class Solution {
public:
    int maxProfit(int k, vector<int>& prices) {
        const int n = prices.size();
        std::unordered_map<std::string, int> cache = {};

        std::function<int(int, int, bool)> transact;
        transact = [&](int date, int limit, bool is_buying){
            if (limit == 0 || date == n)
                return 0;

            std::string state = std::to_string(date) + "," + std::to_string(limit) + "," + std::string{is_buying};

            if (cache.contains(state))
                return cache[state];

            if (is_buying) {
                // either buy today or wait to buy another day
                cache[state] = std::max(
                    transact(date + 1, limit, false) - prices[date],
                    transact(date + 1, limit, true)
                );
            } else {
                cache[state] = std::max(
                    transact(date + 1, limit - 1, true) + prices[date],
                    transact(date + 1, limit, false)
                );
            }

            return cache[state];
        };

        return transact(0, k, true);
    }
};